doc/specs/#7335 - Console Allocation Policy.md

```diff
@@ -0,0 +1,192 @@
+
+Fusion manifest entries are used to make application-scoped decisions like this all the time, like `longPathAware` and
+`heapType`.
+
+CUI applications that can spawn a UI (or GUI applications that can print to a shell) are commonplace on other platforms,
+because there is no subsystem differentiation.
+
+## UI/UX Design
+
+There is no UI for this feature.
+
+## Capabilities
+
+### Accessibility
+
+This should have no impact on accessibility.
+
+### Security
+
+This should have no impact on security.
+
+### Reliability
+
+This should have no impact on reliability.
+
+### Compatibility
+
+On downlevel versions of Windows that do not understand (or expect) this manifest field, applications will launch
+consoles as specified by their image subsystem (described in the [abstract](#abstract) above).
+
+This **will** constitute a breaking change for any application that opts into a change in behavior, but that breaking
+change is expected to be managed by the application.
+
+EXAMPLE: If Python migrates python.exe to specify an allocation policy of **inheritOnly**, graphical python applications
+will become double-click runnable from the graphical shell without spawning a console window. _However_, console-based
+python applications will no longer spawn a console window when double-clicked from the graphical shell.
+
+Python could work around this by calling [`AllocConsole`] if it can be detected that console I/O is required.
+
+### Performance, Power, and Efficiency
+
+This should have no impact on performance, power or efficiency.
+
+## Potential Issues
+
+### Shell Hang
+
+I am **not** proposing a change in how shells determine whether to wait for an application before returning to a prompt.
+This means that a console subsystem application that intends to primarily present a UI but occasionally print text to a
+console (therefore choosing the **inheritOnly** allocation policy) will cause the shell to "hang" and wait for it to
+exit.
+
+Because the vast majority of shells on Windows "hang" by calling `WaitFor...Object` with a HANDLE to the spawned
+process, an application that wants to be a "hybrid" CUI/GUI application will be forced to spawn a separate process to
+detach from the shell and then terminate its main process.
+
+This is very similar to the forking model seen in many POSIX-compliant operating systems.
+
+### Launching interactively from Explorer
+
+Applications like PowerShell may wish to retain "automatic" console allocation, and **inheritOnly** would be unsuitable
+for them. If PowerShell becomes **inheritOnly**, double-clicking `pwsh.exe` will no longer spawn a console. This would
+be a critical failure of the console subsystem.
+
+At the same time, PowerShell wants `-WindowStyle Hidden` to suppress the console _before it's created_.
+
+PowerShell, and any other shell that wishes to maintain interactive launch from the graphical shell, will need to call
+[`AllocConsole`] if it determines that it wants to launch in interactive mode.

```

Comment:
Do you have a theoretical proposal as to how interpreters might determine whether or not they need a console now that it's their problem to discern? Or do you want to leave that completely open ended and up to them?

Reply:
So, PowerShell would likely just `AllocConsole` every time they didn't have a "hidden" window. Because AllocConsole is documented as failing if there's already a console, it's a no-op if they inherited one . . . i think.